#! /bin/sh -e

# All lines beginning with `# DPATCH:' are a description of the patch.
# DP: gcc/config/m68k/m68k.c:
# DP: correctly save the pic register, when not done by reload()
# DP: (fixes _Unwind_RaiseException and thus exception handling)
# DP: fixes 345574

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p1 < $0
        #cd ${dir}gcc && autoconf2.59
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p1 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.

---
 gcc/config/m68k/m68k.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: gcc-4.1/gcc/config/m68k/m68k.c
===================================================================
--- gcc-4.1.orig/gcc/config/m68k/m68k.c
+++ gcc-4.1/gcc/config/m68k/m68k.c
@@ -466,7 +466,8 @@ m68k_save_reg (unsigned int regno, bool 
 {
   if (flag_pic && regno == PIC_OFFSET_TABLE_REGNUM)
     {
-      if (current_function_uses_pic_offset_table)
+      if (current_function_uses_pic_offset_table ||
+	  current_function_has_nonlocal_label)
 	return true;
       if (!current_function_is_leaf && TARGET_ID_SHARED_LIBRARY)
 	return true;
